lunatrace/bsl/backend/src/workers/queue.ts

```diff
@@ -0,0 +1,190 @@
+
+import { sqsClient } from '../aws/sqs-client';
+import { getQueueHandlerConfig, getWorkerBucketConfig } from '../config';
+import { SqsQueueConfig, WorkerBucketConfig } from '../types/config';
+import { MessageTypeToActivityLookup } from '../types/sqs';
+import { MaybeErrorVoid } from '../types/util';
+import { newError, newResult } from '../utils/errors';
+import { log } from '../utils/log';
+import { getSqsUrlFromName } from '../utils/sqs';
+import { catchError, threwError } from '../utils/try';
+
+import {
+  createMessageTypeToActivityLookup,
+  processLunaTraceSqsMessage,
+} from './queue-processors/process-luna-trace-sqs-message';
+import { processS3SqsMessage } from './queue-processors/process-s3-sqs-message';
+
+async function loadQueueUrlFromName(queueName: string): Promise<string> {
+  const queueUrl = await catchError(getSqsUrlFromName(sqsClient, queueName));
+
+  if (threwError(queueUrl) || queueUrl.error) {
+    log.error('unable to load queue url', {
+      queueName: queueName,
+    });
+    process.exit(-1);
+  }
+
+  const loadedQueueUrl = queueUrl.res;
+
+  log.info('Resolved url for queue', {
+    queueName: queueName,
+    loadedQueueUrl,
+  });
+  return loadedQueueUrl;
+}
+
+export async function startQueueWorker(): Promise<void> {
+  const queueWorker = new QueueWorker();
+  await queueWorker.setup();
+
+  await queueWorker.listenForMessages();
+}
+
+class QueueWorker {
+  private queueUrl?: string;
+  private queueConfig: SqsQueueConfig;
+  private activityLookup: MessageTypeToActivityLookup | null;
+  private bucketConfig: WorkerBucketConfig;
+
+  constructor() {
+    this.queueConfig = getQueueHandlerConfig();
+    this.activityLookup = null;
+    this.bucketConfig = getWorkerBucketConfig();
+  }
+
+  async setup() {
+    this.queueUrl = await loadQueueUrlFromName(this.queueConfig.queueName);
+
+    this.activityLookup = await createMessageTypeToActivityLookup();
+  }
+
+  async deleteMessage(message: Message, queueUrl: string) {
+    const deleteParams = {
+      QueueUrl: queueUrl,
+      ReceiptHandle: message.ReceiptHandle,
+    };
+    try {
+      const data = await sqsClient.send(new DeleteMessageCommand(deleteParams));
+      log.info('Message deleted', data);
+    } catch (err) {
+      log.error('Error deleting message', err);
+    }
+  }
+
+  // processMessageAndReturnResults determines how to process a message from a queue
+  async processMessageAndReturnResults(parsedMessage: any): Promise<MaybeErrorVoid[]> {
+    if ('type' in parsedMessage) {
+      // if the key 'type' is present in the message, then this message is one that we have sent from one of our services
+      if (this.activityLookup === null) {
+        throw new Error('activityLookup is null. Make sure that setup was called first.');
+      }
+      return await processLunaTraceSqsMessage(this.activityLookup, parsedMessage);
+    } else if ('Records' in parsedMessage) {
```

Comment:
I guess thats all they really give us. The records have "eventSource":"aws:s3",
But its in the array

Reply:
ill add a TODO in the code that if someone needs to process other aws data sources in this queue that they will need to check the event source.